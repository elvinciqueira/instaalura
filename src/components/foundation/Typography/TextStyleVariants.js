import {css} from 'styled-components'

const paragraph1 = css`
  ${({theme}) => css`
    font-size: ${theme.typographyVariants.paragraph1.fontSize};
    font-weight: ${theme.typographyVariants.paragraph1.fontWeight};
    line-height: ${theme.typographyVariants.paragraph1.lineHeight};
  `}
`

const paragraph2 = css`
  ${({theme}) => css`
    font-size: ${theme.typographyVariants.paragraph2.fontSize};
    font-weight: ${theme.typographyVariants.paragraph2.fontWeight};
    line-height: ${theme.typographyVariants.paragraph2.lineHeight};
  `}
`

const subTitle = css`
  ${({theme}) => css`
    font-size: ${theme.typographyVariants.subTitle.fontSize};
    font-weight: ${theme.typographyVariants.subTitle.fontWeight};
    line-height: ${theme.typographyVariants.subTitle.lineHeight};
  `}
`

const smallestException = css`
  ${({theme}) => css`
    font-size: ${theme.typographyVariants.smallestException.fontSize};
    font-weight: ${theme.typographyVariants.smallestException.fontWeight};
    line-height: ${theme.typographyVariants.smallestException.lineHeight};
  `}
`

const title = css`
  ${({theme}) => css`
    font-size: ${theme.typographyVariants.title.fontSize};
    font-weight: ${theme.typographyVariants.title.fontWeight};
    line-height: ${theme.typographyVariants.title.lineHeight};
  `}
`

const titleXS = css`
  ${({theme}) => css`
    font-size: ${theme.typographyVariants.titleXS.fontSize};
    font-weight: ${theme.typographyVariants.titleXS.fontWeight};
    line-height: ${theme.typographyVariants.titleXS.lineHeight};
  `}
`

export const TextStyleVariants = {
  smallestException,
  paragraph1,
  paragraph2,
  title,
  titleXS,
  subTitle,
}
