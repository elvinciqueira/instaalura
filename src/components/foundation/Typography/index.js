import React from 'react'
import styled, {css} from 'styled-components'
import PropTypes from 'prop-types'
import get from 'lodash/get'

const paragraph1 = css`
  ${({theme}) => css`
    font-size: ${theme.typographyVariants.paragraph1.fontSize};
    font-weight: ${theme.typographyVariants.paragraph1.fontWeight};
    line-height: ${theme.typographyVariants.paragraph1.lineHeight};
  `}
`

const smallestException = css`
  ${({theme}) => css`
    font-size: ${theme.typographyVariants.smallestException.fontSize};
    font-weight: ${theme.typographyVariants.smallestException.fontWeight};
    line-height: ${theme.typographyVariants.smallestException.lineHeight};
  `}
`

export const TextStyleVariants = {
  smallestException,
  paragraph1,
}

const TextBase = styled.span`
  ${({variant}) => TextStyleVariants[variant]}
  color: ${({theme, color}) => get(theme, `colors.${color}.color`)};
`

export default function Typography({tag, variant, children, ...rest}) {
  return (
    <TextBase as={tag} variant={variant} {...rest}>
      {children}
    </TextBase>
  )
}

Typography.defaultProps = {
  tag: 'span',
  variant: 'paragraph1',
}

Typography.propTypes = {
  children: PropTypes.node.isRequired,
  tag: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'p', 'li', 'a', 'span']),
  variant: PropTypes.oneOf(['paragraph1', 'smallestException']),
}
