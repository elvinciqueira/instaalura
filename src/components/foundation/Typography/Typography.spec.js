import React from 'react'
import {render, screen} from '../../../infra/test/testUtils'
import Typography from './index'

describe('<Typography />', () => {
  describe('text variants', () => {
    describe('when variant="title"', () => {
      test('renders with proper font-size', () => {
        const tree = rendererCreateWithTheme(
          <Typography variant="title" />,
        ).toJSON()

        expect(tree).toMatchSnapshot()
        expect(tree).toHaveStyleRule('font-size', '32px')
        expect(tree).toHaveStyleRule('font-weight', '700')
        expect(tree).toHaveStyleRule('line-height', '1.25')
      })
    })

    describe('when passing a href prop', () => {
      test('it renders as a anchor element', () => {
        render(<Typography href="/">LINK</Typography>)

        const link = screen.getByRole('link', {name: /link/i})

        expect(link).toBeInstanceOf(HTMLAnchorElement)
        expect(link).toBeInTheDocument()
      })
    })

    describe('when compoenet renders without variant props', () => {
      test('it applies default styles', () => {
        const tree = rendererCreateWithTheme(<Typography />).toJSON()

        expect(tree).toMatchSnapshot()
        expect(tree).toHaveStyleRule('font-size', '16px')
        expect(tree).toHaveStyleRule('font-weight', '400')
        expect(tree).toHaveStyleRule('line-height', '1.25')
      })
    })
  })

  test('textAlign props sets text-align', () => {
    const tree = rendererCreateWithTheme(
      <Typography textAlign="center" />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('text-align', 'center')
  })

  test('marginTop, marginRight, marginBottom, marginLeft', () => {
    const tree = rendererCreateWithTheme(
      <Typography
        marginTop="10px"
        marginLeft="20px"
        marginRight="20px"
        marginBottom="10px"
      />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('margin-top', '10px')
    expect(tree).toHaveStyleRule('margin-left', '20px')
    expect(tree).toHaveStyleRule('margin-right', '20px')
    expect(tree).toHaveStyleRule('margin-bottom', '10px')
  })
})
