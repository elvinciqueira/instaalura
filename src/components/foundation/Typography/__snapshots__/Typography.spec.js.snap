// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography /> marginTop, marginRight, marginBottom, marginLeft 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
  margin-top: 10px;
  margin-bottom: 10px;
  margin-right: 20px;
  margin-left: 20px;
}

<span
  className="c0"
  href=""
/>
`;

exports[`<Typography /> text variants when compoenet renders without variant props it applies default styles 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
}

<span
  className="c0"
  href=""
/>
`;

exports[`<Typography /> text variants when variant="title" renders with proper font-size 1`] = `
.c0 {
  font-size: 32px;
  font-weight: 700;
  line-height: 1.25;
}

<span
  className="c0"
  href=""
/>
`;

exports[`<Typography /> textAlign props sets text-align 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
  text-align: center;
}

<span
  className="c0"
  href=""
/>
`;
